File: 백준/Gold/15486. 퇴사 2/퇴사 2.js
const fs = require('fs');
const stdin = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

// N이 150만이라서, O(N) 이 필요한 듯! => dp
// N + 1 이 되는 날에 퇴사한다.
const n = parseInt(stdin[0]);

const [t, p] = stdin
  .slice(1)
  .map((row) => row.split(' ').map(Number))
  .reduce(
    (acc, [t_i, p_i]) => {
      acc[0].push(t_i);
      acc[1].push(p_i);

      return acc;
    },
    [[], []],
  );

// dp[i] : i일차에 상담을 하면 얻을 수 있는 최대 이익
const dp = Array.from({ length: n + 1 }, () => 0);

// init
dp[n - 1] = n - 1 + t[n - 1] > n ? 0 : p[n - 1];

for (let i = n - 2; i >= 0; i--) {
  const next = i + t[i];

  // 해당 날짜에 상담을 할 수 있는 경우
  if (next <= n) {
    dp[i] = Math.max(dp[i + 1], p[i] + dp[next]);
  } else {
    dp[i] = dp[i + 1];
  }
}

console.log(dp[0]);

File: 백준/Gold/15486. 퇴사 2/퇴사 2.java
import java.io.*;
import java.util.Arrays;

public class Main {
    static int n;
    static int[] t, p;

    // dp[i] : i일에 상담을 했을 때, 얻을 수 있는 최대 이익
    static int[] dp;

    public static void main(String[] args) throws IOException {
        input();

        dp = new int[n + 1];
        dp[n - 1] = n - 1 + t[n - 1] > n ? 0 : p[n - 1];

        for (int i = n - 2; i >= 0; i--){
            int next = i + t[i];
            dp[i] = dp[i + 1];

            if (next <= n){
                dp[i] = Math.max(dp[i + 1], p[i] + dp[next]);
            }
        }

        System.out.println(dp[0]);
    }

    public static void input() throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());
        t = new int[n];
        p = new int[n];

        for (int i = 0; i < n; i++){
            int[] splited = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
            t[i] = splited[0];
            p[i]= splited[1];
        }
    }
}
